oxide-auth/src/code_grant/extensions/pkce.rs

```diff
@@ -1,11 +1,46 @@
 use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
 
-use crate::primitives::grant::{GrantExtension, Value};
+use crate::{
+    primitives::grant::{GrantExtension, Value},
+    util::avoid_alloc_cow_str_to_string,
+};
 
 use base64;
 use sha2::{Digest, Sha256};
 use subtle::ConstantTimeEq;
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PkceError {
+    NoChallenge,
+    NoVerifier,
+    FailedVerification(&'static str),
+    InvalidMethod(String),
+    UnsupportedMethod,
+}
+
+impl Display for PkceError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                // perhaps use thiserror so we can avoid these allocs?
+                PkceError::NoChallenge =>
+                    "PKCE: It was indicated that an error was required but none was given".to_string(),
+                PkceError::NoVerifier =>
+                    "PKCE: The challenge was agreed upon but there is no verifier present".to_string(),
+                PkceError::FailedVerification(reason) => format!("PKCE: Failed to verify PKCE method: {reason}"),
+                PkceError::InvalidMethod(method) => format!("PKCE: Method must be either `S256`(sha256) or `plain`, but was given: {method}"),
+                PkceError::UnsupportedMethod => "PKCE: Method is unsupported due to indicator `allow_plain: false` and `method: plain`".to_string(),
```

Comment:
Then the variant should be named `UnsupportedMethodPlain`. Also, don't refer to server internals in user-facing errors. This is fine as a Debug message but less so as Display.